v5_MIGRATION_GUIDE.md

```diff
@@ -19,5 +20,46 @@ Now, you can tokenize with just the card details:
 
 // TODO: code snippet of card integration in v5
 
+## Venmo
+
+The Venmo integration has been updated to allow for more flexibility with app switching and 
+activity result handling. 
+
+`VenmoLauncher` has been added to handle the app switching portion of the Venmo flow for user 
+authentication via the Venmo app. This class uses the Android Activity Result API and therefore 
+must be instantiated in the `OnCreate` method of your Activity or Fragment.
+
+`BraintreeClient` and `VenmoClient` no longer require references to Fragment or Activity and do not 
+need to be instantiated in `OnCreate`.
+
+```kotlin
 
+class MyActivity : FragmentActivity() {
+    
+    private lateinit var venmoLauncher: VenmoLauncher
+    private lateinit var braintreeClient: BraintreeClient
+    private lateinit var venmoClient: VenmoClient
+    
+    @override fun onCreate(savedInstanceState: Bundle?) {
+        venmoLauncher = VenmoLauncher(this) { authChallengeResult ->
```

Comment:
I think of "authentication challenge" as a barrier to protect against unauthorized users, and less so for general login. According to[ this definition](https://en.wikipedia.org/wiki/Challenge%E2%80%93response_authentication) though, a basic password authentication does qualify.

Just food for thought that this language might be confusing since for flows (3DS) the "challenge" is optional and not required.

Reply:
Maybe `continue` or `presentLogin` or `startLogin`?

Reply:
@sshropshire what are your thoughts on changing up this naming scheme? 

I'd be fine with something like `venmoClient.presentLogin` returning a `LoginResult` or keeping with this `authChallenge` naming. Thinking through the other modules if we want consistent naming:

PayPal and Venmo are "login" screens
Google Pay is just the Google Pay dialog so not sure either naming makes sense for that one (same with PayPalNativeCheckout)
3DS is a true auth challenge
Local Payment is likely a "login" screen
SEPA launches a mandate

Maybe it won't make sense to have naming consistency and we should just describe what each flow is doing? I do like the clarity across integrations of a single pattern though

Reply:
Yeh, I'm with you on preferring a unified name across modules!

Reply:
I have a question which is _slightly tangential_, but I think could inform the discussion:

For 3DS, we'll still require merchants to call [`continuePerformVerification`](https://github.com/braintree/braintree_android/blob/d7e67f8bf1aa25a7560f4e2975543a0f33533778/ThreeDSecure/src/main/java/com/braintreepayments/api/ThreeDSecureClient.java#L265) in order to proceed with the challenge?

So would that look something like this (in pseudocode/short-hand)?

```
class MyActivity {
    fun onCreate() {
        threeDSecureLauncher { authChallengeResult ->
            threeDSecureClient.performVerification() {
                threeDSecureClient. continuePerformVerification()
            }
        }
    }

    func onClick() {
        threeDSecureClient.requestAuthChallenge()
    }
}

class MyClass: ThreeDSecureListener {
    fun onThreeDSecureSuccess() {
        // handle result/nonce
    }
}
```

Would the launcher & client steps be flipped?

Reply:
In the 3DS case, the `performVerification` method becomes `requestAuthChallenge` (or we could leave the `performVerification` naming as is)

```
class MyActivity {
    fun onCreate() {
        threeDSecureLauncher = ThreeDSecureLauncher(this) { authChallengeResult ->
              threeDSecureClient.continuePerformVerification() { nonce, error -> 
                  // handle nonce/error
             }
        }
    }

    func onClick() {
        threeDSecureClient.requestAuthChallenge(this, request) { authChallenge, error -> 
          threeDSecureLauncher.launch(authChallenge)
        }
    }
}
```

Reply:
I've added a ticket to re-asses the naming of the `requestAuthChallenge` method and make sure it makes sense across modules so I'm going to leave it as is in this PR

Reply:
I'm all for merchants having an intuitive interface. It's a good callout to try and improve naming whenever we can! 

I don't have a major preference, but with this new design pattern the `Launcher` will be presenting the login flow instead of the `Client`. We talked about potentially allowing the Launcher to have a reference to the Client, that would make methods like `launcher.startLogin()` possible.

The current naming is tightly coupled to the underlying Android frameworks we use. The way it is now helps us a lot more than it'll help merchants. We should iterate on naming before GA to maximize API clarity.